### PR TITLE
fix: Broken cache on Windows

### PR DESCRIPTION
**Pull request by serverwentdown**
(Backport to 1.1: #4549)

The previous implementation would fail to install packages on Windows because it creates a `Path` starting with a slash. Such a `Path` is invalid on Windows. Instead, use the utility function url_to_path.

Additionally (not a bug in 1.1, migrated from separate PR at #4532), Links and Paths were passed in to pip_install as a string. This makes it hard for pip_install to parse out the extension of the file to install. By accepting only Links or Paths, pip_install can parse out the extension using the Link. This also resolves a bug when Executor._download_link returns a Link, which can't be parsed by Path(str(link)) on Windows.

This was overlooked when running on Linux because file:/path/to/cache is somehow a valid pip install path in Python on Linux, but not a valid path on Windows. With this PR, pip install is always called with a proper Path (/path/to/cache) and never a Path-parsed Link (file:/path/to/cache)

# Pull Request Check List

Resolves: #4479 #4535

- [x] Added **tests** for changed code
- [x] Updated **documentation** for changed code. (N/A)

### CODE REVIEW

**Comment by Ivoz** [comment]
I wonder if we can create a testcase that will fail that this fixes?

**Comment by serverwentdown** [comment]
@Ivoz do you know where to put full end-to-end tests (running poetry commands)?

**Comment by mrogaski** [comment]
I confirmed that this fixes #4541.

**Comment by serverwentdown** [comment]
@finswimmer @sdispater I'd like a review of this PR because this is a major issue for all Windows users

**Comment by serverwentdown** [comment]
@Ivoz Oh you're right! Thanks for catching this lack of test coverage. I'll fix it after I sleep, but my main suspect is that fixture_dir returns a relative path hence the bug does not get triggered.

**Review by abn — DISMISSED** [review]
Overall looks good. Some minor suggestions.

**Comment by serverwentdown** [comment]
I have made changes a while back, need another review.
@abn? @sdispater?

File: poetry/utils/pip.py
Comment (neersighted) [reply poetry/utils/pip.py:16]: I have to agree with @abn here -- that function is not long for the world, but for now, we'd like it to match -- when it's removed, the method of excision will be more obvious.

**Review by neersighted — CHANGES_REQUESTED** [review]
This looks a lot better than the current code -- we're making a lot better use of duck typing while eliminating a rather silly corner case. There's still some `isinstance()`, but I think you've got it down to the bare minimum

I'm ready to merge this once it passes lint and is rebased onto the current `master` (the FreeBSD tests were fixed there) -- and please make sure to make the same post-review changes to your backport PR.

**Comment by neersighted** [comment]
Looks like this still needs to be rebased to fix the merge conflicts and fix the failing FreeBSD tests. I'll review after that's complete.

**Review by neersighted — APPROVED** [review]

**Comment by neersighted** [comment]
Merged! Thank you for your work on this -- and sorry that poetry has been broken so long!

**Comment by serverwentdown** [comment]
Awesome, thanks for getting this finally merged! 🥳

**Comment by mario-bermonti** [comment]
Great! Thanks for fixing this.

**Pull request merged**
